modules/install/src/main/scala/coursier/install/RawAppDescriptor.scala

```diff
@@ -216,6 +246,71 @@ object RawAppDescriptor {
 
   }
 
+  @data class RawVersionOverride(
+    versionRange: String,
+    dependencies: Option[List[String]] = None,
+    repositories: Option[List[String]] = None,
+    mainClass: Option[String] = None,
+    properties: Option[RawAppDescriptor.Properties] = None
+  ) {
+    def versionOverride: ValidatedNel[String, VersionOverride] = {
+      val versionRangeV = coursier.core.Parse
+        .versionInterval(versionRange)
+        .toValidNel(s"""versionRange "$versionRange" is invalid""")
+      val repositoriesV = repositories.map(parseRepositories).sequence
+      val dependenciesV = dependencies.map(parseDependenices).sequence
+      val (mainClassOpt, defaultMainClassOpt) = mainClass.map(parseMainClass) match {
+        case Some(Left(mainClass))         => (Some(mainClass), Some(""))
+        case Some(Right(defaultMainClass)) => (Some(""), Some(defaultMainClass))
+        case None                          => (None, None)
+      }
+
+      (versionRangeV, repositoriesV, dependenciesV).mapN {
+        (versionRange, repositories, dependencies) =>
+          VersionOverride(versionRange)
+            .withDependencies(dependencies)
+            .withRepositories(repositories)
+            .withMainClass(mainClassOpt)
+            .withDefaultMainClass(defaultMainClassOpt)
+            .withJavaProperties(properties.map(_.props.sorted))
+      }
+    }
+  }
+
+  /* Left is mainClass and Right is defaultMainClass */
+  private def parseMainClass(mainClass: String): Either[String, String] =
+    if (mainClass.endsWith("?")) Right(mainClass.stripSuffix("?"))
+    else Left(mainClass)
+
+  private def parseDependenices(dependencies: Seq[String])
+    : ValidatedNel[String, Seq[JavaOrScalaDependency]] =
+    validationNelToCats(
+      DependencyParser.javaOrScalaDependencies(dependencies, Configuration.defaultCompile)
+    )
+
+  private def parseRepositories(repositories: Seq[String]): ValidatedNel[String, Seq[Repository]] =
+    validationNelToCats(RepositoryParser.repositories(repositories))
+
+  /** Check that there is no overlapping between version intervals
+    */
+  private[install] def validateRanges(versionOverrides: Seq[VersionOverride])
+    : ValidatedNel[String, Seq[VersionOverride]] =
+    versionOverrides
+      .map(_.versionRange)
+      .foldLeft[ValidatedNel[String, Seq[VersionInterval]]](Validated.valid(Seq.empty)) {
+        case (validRanges, range) =>
+          validRanges.andThen { ranges =>
+            val conflictingRanges = ranges.filter(_.merge(range).nonEmpty)
+
+            if (conflictingRanges.isEmpty) Validated.valid(ranges :+ range)
+            else {
+              val conflicts = conflictingRanges.map(i => "\"$i\"").mkString("[", ", ", "]")
```

Comment:
This isn't interpolating.

Reply:
Should be fixed by https://github.com/coursier/coursier/pull/2513.